File: Submissions/LeetCode/611. Valid Triangle Number.py
class Solution:
    def fun(self, another, sum):
        l = 0
        r = len(another) - 1
        ans = len(another)
        while (l <= r) :
            mid = int((l + r) / 2)
            if another[mid] >= sum:
                l = mid + 1
            else:
                ans = mid
                r = mid - 1
        
        return (len(another) - ans)

    def triangleNumber(self, nums: list[int]) -> int:
        another = []
        nums.sort()
        ans = 0
        for i in range(len(nums) - 1, 0, -1):
            for j in range(i - 1, -1, -1):
                sum = nums[i] + nums[j]
                ans = ans + self.fun(another, sum)
            another.append(nums[i])
        
        return ans

nums = [4,2,3,4]
c = Solution()
print(c.triangleNumber(nums))

File: Submissions/LeetCode/611. Valid Triangle Number.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  int triangleNumber(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    int n = (int)nums.size();
    deque<int> dq;
    int ans = 0;
    for (int i = n - 1; i >= 0; i--) {
      for (int j = i - 1; j >= 0; j--) {
        int sum = nums[i] + nums[j];
        ans += lower_bound(dq.begin(), dq.end(), sum) - dq.begin();
      }
      dq.push_front(nums[i]);
    }
    return ans;
  }
};

int32_t main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int n; cin >> n;
  vector<int> nums(n);
  for (auto &x : nums) {
    cin >> x;
  }
  Solution c;
  cout << c.triangleNumber(nums);

  return 0;
}